File: sdk/typescript/src/tools/web-search.ts
/**
 * Web search tool — lets agents search the web for current information.
 *
 * Defaults to the Tavily Search API using raw fetch() (no additional
 * dependencies). Users can plug in any search provider via a custom
 * function.
 *
 * @example
 * ```typescript
 * import { createWebSearchTool } from '@polos/sdk';
 *
 * // Tavily with env var (TAVILY_API_KEY)
 * const webSearch = createWebSearchTool();
 *
 * // Custom provider
 * const webSearch = createWebSearchTool({
 *   search: async (query, opts) => {
 *     const res = await mySearchApi(query, opts.maxResults);
 *     return { query, results: res.items };
 *   },
 * });
 * ```
 */

import { z } from 'zod';
import { defineTool } from '../core/tool.js';
import type { ToolWorkflow } from '../core/tool.js';

// ── Result types (provider-agnostic) ─────────────────────────────────

/** A single search result item. */
export interface WebSearchResultItem {
  title: string;
  url: string;
  /** Snippet or summary of the page content. */
  content: string;
  /** Relevance score, 0–1. */
  score?: number | undefined;
  /** Publication date in ISO 8601 format. */
  publishedDate?: string | undefined;
}

/** Full search result returned by the tool. */
export interface WebSearchResult {
  query: string;
  results: WebSearchResultItem[];
  /** AI-generated summary (Tavily feature). */
  answer?: string | undefined;
}

// ── Options passed to the search function ────────────────────────────

/** Options forwarded to the search function at call time. */
export interface WebSearchOptions {
  maxResults: number;
  topic: 'general' | 'news';
}

// ── Custom provider interface ────────────────────────────────────────

/** Signature for a custom search provider function. */
export type WebSearchFunction = (
  query: string,
  options: WebSearchOptions
) => Promise<WebSearchResult>;

// ── Configuration ────────────────────────────────────────────────────

/** Tavily-specific configuration knobs. */
export interface TavilySearchConfig {
  /** Tavily API key. Falls back to the TAVILY_API_KEY environment variable. */
  apiKey?: string;
  /** Search depth. @default 'basic' */
  searchDepth?: 'basic' | 'advanced';
  /** Include an AI-generated answer in the response. @default true */
  includeAnswer?: boolean;
  /** Include raw page content in results. @default false */
  includeRawContent?: boolean;
  /** Tavily API base URL. @default 'https://api.tavily.com' */
  baseUrl?: string;
}

/** Configuration for createWebSearchTool(). */
export interface WebSearchToolConfig extends TavilySearchConfig {
  /** Custom search provider. When set, overrides the built-in Tavily implementation. */
  search?: WebSearchFunction;
  /** Default maximum results per query. @default 5 */
  maxResults?: number;
  /** Default topic filter. @default 'general' */
  topic?: 'general' | 'news';
  /** Tool identifier exposed to the LLM. @default 'web_search' */
  toolId?: string;
}

// ── LLM-facing input schema ──────────────────────────────────────────

const webSearchInputSchema = z.object({
  query: z.string().describe('The search query'),
  maxResults: z.number().optional().describe('Maximum number of results to return'),
  topic: z
    .enum(['general', 'news'])
    .optional()
    .describe('Topic filter: general web search or news'),
});

type WebSearchInput = z.infer<typeof webSearchInputSchema>;

// ── Tavily implementation (internal) ─────────────────────────────────

interface TavilyResponse {
  query: string;
  answer?: string;
  results: {
    title: string;
    url: string;
    content: string;
    score?: number;
    published_date?: string;
    raw_content?: string;
  }[];
}

function createTavilySearchFn(config: TavilySearchConfig): WebSearchFunction {
  return async (query: string, options: WebSearchOptions): Promise<WebSearchResult> => {
    const apiKey = config.apiKey ?? process.env['TAVILY_API_KEY'];
    if (!apiKey) {
      throw new Error(
        'Tavily API key is required. Provide it via the apiKey option or set the TAVILY_API_KEY environment variable.'
      );
    }

    const baseUrl = (config.baseUrl ?? 'https://api.tavily.com').replace(/\/+$/, '');

    const body = {
      query,
      max_results: options.maxResults,
      search_depth: config.searchDepth ?? 'basic',
      include_answer: config.includeAnswer ?? true,
      include_raw_content: config.includeRawContent ?? false,
      topic: options.topic,
    };

    const response = await fetch(`${baseUrl}/search`, {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify(body),
    });

    if (!response.ok) {
      let errorMessage: string;
      try {
        const errorBody = (await response.json()) as Record<string, unknown>;
        errorMessage =
          typeof errorBody['detail'] === 'string' ? errorBody['detail'] : JSON.stringify(errorBody);
      } catch {
        errorMessage = await response.text();
      }
      throw new Error(`Tavily API error (${String(response.status)}): ${errorMessage}`);
    }

    const data = (await response.json()) as TavilyResponse;

    return {
      query: data.query,
      answer: data.answer,
      results: data.results.map((r) => ({
        title: r.title,
        url: r.url,
        content: r.content,
        score: r.score,
        publishedDate: r.published_date,
      })),
    };
  };
}

// ── Factory ──────────────────────────────────────────────────────────

/**
 * Create a web search tool for agent use.
 *
 * By default uses the Tavily Search API via raw fetch(). Pass a custom
 * `search` function to use any other provider.
 *
 * @example
 * ```typescript
 * // Tavily with env var
 * const webSearch = createWebSearchTool();
 *
 * // Tavily with explicit key
 * const webSearch = createWebSearchTool({ apiKey: 'tvly-xxx' });
 *
 * // Custom provider
 * const webSearch = createWebSearchTool({
 *   search: async (query, opts) => {
 *     const res = await myApi(query, opts.maxResults);
 *     return { query, results: res.items };
 *   },
 * });
 * ```
 */
export function createWebSearchTool(config?: WebSearchToolConfig): ToolWorkflow {
  const toolId = config?.toolId ?? 'web_search';
  const defaultMaxResults = config?.maxResults ?? 5;
  const defaultTopic = config?.topic ?? 'general';

  // Search function resolved lazily — custom provider or Tavily default.
  // For Tavily, the API key is resolved at call time (not factory time).
  const searchFn: WebSearchFunction = config?.search ?? createTavilySearchFn(config ?? {});

  return defineTool(
    {
      id: toolId,
      description:
        'Search the web for current information. ' +
        'Returns a list of relevant results with titles, URLs, and content snippets.',
      inputSchema: webSearchInputSchema,
    },
    async (ctx, input: WebSearchInput) => {
      const options: WebSearchOptions = {
        maxResults: input.maxResults ?? defaultMaxResults,
        topic: input.topic ?? defaultTopic,
      };

      const result = await ctx.step.run('web_search', () => searchFn(input.query, options), {
        input: { query: input.query, options },
      });

      return result;
    }
  ) as ToolWorkflow;
}

File: sdk/python/polos/tools/web_search.py
"""Web search tool -- lets agents search the web for current information.

Defaults to the Tavily Search API using httpx (no additional
dependencies beyond the SDK's existing httpx requirement).  Users can
plug in any search provider via a custom async function.

Example::

    from polos import create_web_search_tool

    # Tavily with env var (TAVILY_API_KEY)
    web_search = create_web_search_tool()

    # Custom provider
    web_search = create_web_search_tool(WebSearchToolConfig(
        search=my_custom_search_fn,
    ))
"""

from __future__ import annotations

import os
from collections.abc import Awaitable, Callable
from dataclasses import dataclass
from typing import Any, Literal

import httpx
from pydantic import BaseModel, Field

from ..core.context import WorkflowContext
from .tool import Tool, ToolApproval

# ── Result types (provider-agnostic) ──────────────────────────────────


class WebSearchResultItem(BaseModel):
    """A single search result item."""

    title: str
    url: str
    content: str
    """Snippet or summary of the page content."""
    score: float | None = None
    """Relevance score, 0-1."""
    published_date: str | None = None
    """Publication date in ISO 8601 format."""


class WebSearchResult(BaseModel):
    """Full search result returned by the tool."""

    query: str
    results: list[WebSearchResultItem]
    answer: str | None = None
    """AI-generated summary (Tavily feature)."""


# ── Options passed to the search function ─────────────────────────────


class WebSearchOptions(BaseModel):
    """Options forwarded to the search function at call time."""

    max_results: int
    topic: Literal["general", "news"]


# ── Custom provider type ──────────────────────────────────────────────

WebSearchFunction = Callable[[str, WebSearchOptions], Awaitable[WebSearchResult]]

# ── Configuration ─────────────────────────────────────────────────────


@dataclass
class TavilySearchConfig:
    """Tavily-specific configuration knobs."""

    api_key: str | None = None
    """Tavily API key.  Falls back to the ``TAVILY_API_KEY`` environment variable."""
    search_depth: Literal["basic", "advanced"] = "basic"
    """Search depth."""
    include_answer: bool = True
    """Include an AI-generated answer in the response."""
    include_raw_content: bool = False
    """Include raw page content in results."""
    base_url: str = "https://api.tavily.com"
    """Tavily API base URL."""


@dataclass
class WebSearchToolConfig(TavilySearchConfig):
    """Configuration for :func:`create_web_search_tool`."""

    search: WebSearchFunction | None = None
    """Custom search provider.  Overrides the built-in Tavily implementation."""
    max_results: int = 5
    """Default maximum results per query."""
    topic: Literal["general", "news"] = "general"
    """Default topic filter."""
    tool_id: str = "web_search"
    """Tool identifier exposed to the LLM."""
    approval: ToolApproval | None = None
    """Require human approval before execution."""


# ── LLM-facing input schema ──────────────────────────────────────────


class WebSearchInput(BaseModel):
    """Input schema for the web_search tool."""

    query: str = Field(description="The search query")
    max_results: int | None = Field(
        default=None,
        alias="maxResults",
        description="Maximum number of results to return",
    )
    topic: Literal["general", "news"] | None = Field(
        default=None, description="Topic filter: general web search or news"
    )

    model_config = {"populate_by_name": True}


# ── Tavily implementation (internal) ──────────────────────────────────


def _create_tavily_search_fn(config: TavilySearchConfig) -> WebSearchFunction:
    """Build an async search function backed by the Tavily Search API."""

    async def _tavily_search(query: str, options: WebSearchOptions) -> WebSearchResult:
        api_key = config.api_key or os.environ.get("TAVILY_API_KEY")
        if not api_key:
            raise RuntimeError(
                "Tavily API key is required. Provide it via the api_key option "
                "or set the TAVILY_API_KEY environment variable."
            )

        base_url = config.base_url.rstrip("/")

        body = {
            "query": query,
            "max_results": options.max_results,
            "search_depth": config.search_depth,
            "include_answer": config.include_answer,
            "include_raw_content": config.include_raw_content,
            "topic": options.topic,
        }

        async with httpx.AsyncClient() as client:
            response = await client.post(
                f"{base_url}/search",
                json=body,
                headers={
                    "Content-Type": "application/json",
                    "Authorization": f"Bearer {api_key}",
                },
            )

        if response.status_code != 200:
            try:
                error_body = response.json()
                error_message = (
                    error_body.get("detail")
                    if isinstance(error_body.get("detail"), str)
                    else response.text
                )
            except Exception:
                error_message = response.text
            raise RuntimeError(f"Tavily API error ({response.status_code}): {error_message}")

        data = response.json()

        return WebSearchResult(
            query=data.get("query", query),
            answer=data.get("answer"),
            results=[
                WebSearchResultItem(
                    title=r["title"],
                    url=r["url"],
                    content=r["content"],
                    score=r.get("score"),
                    published_date=r.get("published_date"),
                )
                for r in data.get("results", [])
            ],
        )

    return _tavily_search


# ── Factory ───────────────────────────────────────────────────────────


def create_web_search_tool(config: WebSearchToolConfig | None = None) -> Tool:
    """Create a web search tool for agent use.

    By default uses the Tavily Search API via httpx.  Pass a custom
    ``search`` function to use any other provider.

    Example::

        from polos import create_web_search_tool, WebSearchToolConfig

        # Tavily with env var
        web_search = create_web_search_tool()

        # Tavily with explicit key
        web_search = create_web_search_tool(WebSearchToolConfig(api_key="tvly-xxx"))

        # Custom provider
        web_search = create_web_search_tool(WebSearchToolConfig(
            search=my_search_fn,
        ))
    """
    config = config or WebSearchToolConfig()
    tool_id = config.tool_id
    default_max_results = config.max_results
    default_topic = config.topic

    # Search function -- custom provider or Tavily default.
    # For Tavily, the API key is resolved at call time (not factory time).
    search_fn: WebSearchFunction = config.search or _create_tavily_search_fn(config)

    async def handler(ctx: WorkflowContext, input: WebSearchInput) -> dict[str, Any]:
        options = WebSearchOptions(
            max_results=input.max_results if input.max_results is not None else default_max_results,
            topic=input.topic or default_topic,
        )

        async def _do_search() -> WebSearchResult:
            return await search_fn(input.query, options)

        result: WebSearchResult = await ctx.step.run("web_search", _do_search)

        return {
            "query": result.query,
            "answer": result.answer,
            "results": [
                {
                    "title": r.title,
                    "url": r.url,
                    "content": r.content,
                    "score": r.score,
                    "publishedDate": r.published_date,
                }
                for r in result.results
            ],
        }

    async def wrapped_func(ctx: WorkflowContext, payload: dict[str, Any] | None):
        input_obj = WebSearchInput.model_validate(payload) if payload else WebSearchInput(query="")
        return await handler(ctx, input_obj)

    # Build JSON schema and remap max_results -> maxResults for LLM
    schema = WebSearchInput.model_json_schema()

    tool = Tool(
        id=tool_id,
        description=(
            "Search the web for current information. "
            "Returns a list of relevant results with titles, URLs, and content snippets."
        ),
        parameters=schema,
        func=wrapped_func,
        approval=config.approval,
    )
    tool._input_schema_class = WebSearchInput
    return tool
